### PR TITLE
[None][feat] Add SiTU support to Rubin FC12 MoE backend

### PR DESCRIPTION
**Pull request by xguannv**
## Dev Engineer Review

- Adds Rubin FC12 support for `ActivationType.SiTu` while preserving `Swiglu` and clamp behavior.
- Propagates activation and SiTU beta values through the operator, runner, generated kernel, and autotuner identities.
- Validates supported activations, SiTU parameters, softcaps, and incompatible clamp settings.
- Adds `CUTEDSL_FC12` as a strict Kimi K3 SiTU backend.
- Review finding counts are unavailable.
- CI follow-up remains required for the Rubin capability skip condition, the merge from `main`, and focused Rubin test reruns.

## QA Engineer Review

- Updates FC12 backend, Kimi K3 SiTU, and activation-resolution tests.
- Covers SiTU softcaps, nonuniform-softcap rejection, SwiGLU clamp behavior, cache-key separation, FC12 preparation, unsupported activations, and CUDA Graph replay.
- Renames the CI test-list entry to `test_codegen_baked_situ_softcaps_are_uniform_scalars`.
- Coverage verdict: **needs follow-up** because CI reported an unstable L0 result and pending Rubin test updates.

## Per-File QA Perspective

- `tensorrt_llm/_torch/custom_ops/cute_dsl_custom_ops.py`: Verify activation and SiTU beta defaults, schema changes, and forwarding.
- `tensorrt_llm/_torch/cute_dsl_kernels/rubin/moe/rubin_contiguous_grouped_blockscaled_gemm_fused_fc12.py`: Verify SiTU epilogues, validation, clamp rejection, and SwiGLU behavior.
- `tensorrt_llm/_torch/moe/fused_moe/fused_moe_cute_dsl_fc12.py`: Verify capability reporting, autotuner-key separation, and FC12 preparation.
- `tensorrt_llm/_torch/models/modeling_kimi_linear.py`: Verify strict Kimi K3 routing to `CUTEDSL_FC12`.
- `tests/integration/test_lists/test-db/l0_b200.yml`: Renames the SiTU softcap test entry in the CI test database.
- `tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py`: Covers Rubin FC12 execution, capability gating, CUDA Graph replay, and backend preservation. CI-list coverage is not established.
- `tests/unittest/_torch/moe/test_moe_backend.py`: Covers softcaps, cache identities, clamp behavior, and FC12 preparation. CI-list coverage is not established.
- `tests/unittest/_torch/moe/test_moe_impl.py`: Covers SiTU acceptance and `Relu2` rejection. CI-list coverage is not established.

## Description

PR #18357 added the Rubin FC12 MoE backend, but its fused epilogue supports only SwiGLU. Kimi K3 uses SiTU, so the resolver cannot select FC12 for those MoE layers.

This change:
- specializes the FC12 epilogue for SwiGLU or SiTU;
- passes the SiTU softcaps through the backend and custom op; and
- includes the activation and softcaps in both compiled-kernel and outer-autotuner identities.

Existing SwiGLU and clamp behavior remains supported.

## Test Coverage

Added tests cover:
- FC12 activation selection and rejection of unsupported activations;
- validation that SiTU softcaps are uniform scalars;
- compiled-kernel and autotuner cache-key separation by activation and softcaps;
- FC12 preparation behavior; and
- a Rubin-only real FC12 SiTU comparison with CUDA Graph replay and a plain-SwiGLU negative reference.

Validation:
- All pre-commit hooks applicable to the seven modified files passed, including both repository-wide test-list checks.
- `python3 -m py_compile` passed for all six modified Python files.
- Hecate SM107, implementation commit `51ec5e46b792ef64fd9e5db96dc49edb09bc7ba8`, Slurm job `621253`: the TensorRT-LLM build with `CMAKE_CUDA_ARCHITECTURES=107-real` passed. The three implementation-file blobs are unchanged in the current rebased commit `666fc3a35459369f9bf23463d99a3f4fa4ecd4d6`.
- FC12 registry and activation tests: 4 passed.
- FC12 backend and NVFP4 shape tests: 16 passed; 10 locality-domain variants skipped by their existing condition.
- Real `CUTEDSL_FC12` SiTU kernel and CUDA Graph test: passed. SiTU reference: cosine `0.972989`, relative L2 `0.232667`. Incorrect SwiGLU reference: cosine `0.065067`, relative L2 `0.998031`.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

**Comment by BowenFu** [comment]
**Blocked for Rubin 1.3 on two mechanical items.**

Please update the FC12 SiTU test skip condition to require explicit `MoED.CUTEDSL_RUBIN` capability while retaining the CUDA/SM107 checks, then merge current `main` to resolve the conflict and rerun the focused Rubin backend/shape and real-kernel tests. The hardware results in the description are useful; once these two items are on the live head, this can return to review.

**Review by BowenFu — CHANGES_REQUESTED** [review]
**One correctness guard must be fixed before re-review.**

The FC12 SiTU test currently gates on CUDA and SM107 but not on the explicit `MoED.CUTEDSL_RUBIN` capability it exercises. Please add that capability requirement while preserving the existing architecture checks, then resolve the `main` conflict and rerun the focused Rubin backend/shape and real-kernel tests. This is required for this PR; otherwise unsupported configurations can enter a capability-specific test path.

**Review by BowenFu — APPROVED** [review]
**Approved on functionality; the bot’s remaining test-guard item is non-blocking.**

The Rubin FC12 SiTU implementation has focused backend/shape coverage and reported real-kernel plus CUDA-graph validation. Adding `MoED.CUTEDSL_RUBIN` to the test skip condition is worthwhile test hygiene, but it does not invalidate the supported SM107 path. The branch still must resolve its `main` conflict before merge.

**Review by leslie-fang25 — APPROVED** [review]

**Review by yuxianq — APPROVED** [review]
No attention change, approve to unblock it

File: tensorrt_llm/_torch/cute_dsl_kernels/rubin/moe/rubin_contiguous_grouped_blockscaled_gemm_fused_fc12.py [hunk 1]
```diff
@@ -937,6 +964,31 @@ def __init__(
...
         self.swiglu_limit = swiglu_limit
         self.has_swiglu_limit = swiglu_limit != float("inf")
+        self.activation_type = validate_fc12_activation_type(activation_type)
+        if self.activation_type == ActivationType.SiTu:
+            if situ_beta is None or situ_linear_beta is None:
+                raise ValueError(
+                    "ActivationType.SiTu requires both situ_beta and "
+                    f"situ_linear_beta, got {situ_beta} and {situ_linear_beta}."
+                )
+            if situ_beta <= 0 or situ_linear_beta <= 0:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/cute_dsl_kernels/rubin/moe/rubin_contiguous_grouped_blockscaled_gemm_fused_fc12.py]: `<= 0` admits `inf` and `nan` (every comparison with nan is False). Both are then folded into the epilogue constants: with `beta=inf`, `2.0/beta`, `2.0*beta`, `-beta` become `0, inf, -inf` and every output element is `inf*sigmoid(0) - inf = NaN`. The Blackwell act-fusion kernel checks `0 < situ_beta < _INF and 0 < situ_linear_beta < _INF` for exactly this reason; please use the same bounded check here (or `math.isfinite`) so a bad softcap raises at construction instead of producing NaNs at runtime.

File: tensorrt_llm/_torch/cute_dsl_kernels/rubin/moe/rubin_contiguous_grouped_blockscaled_gemm_fused_fc12.py [hunk 2]
```diff
@@ -2199,6 +2251,173 @@ class SharedStorageCpasync2cta:
...
+                    acc_vec_gate_alpha = fmin(acc_vec_gate_alpha, self.swiglu_limit)
+                    acc_vec_up_alpha = fclip_xorsign(
+                        acc_vec_up_alpha, self.swiglu_limit
+                    )
+                tCompute[i] = acc_vec_up_alpha * silu_f32(
+                    acc_vec_gate_alpha, fastmath=True
+                )
+
+    @cute.jit
+    def _apply_situ_epilogue(
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/cute_dsl_kernels/rubin/moe/rubin_contiguous_grouped_blockscaled_gemm_fused_fc12.py]: This method and the nested `_sigmoid` are byte-for-byte copies of `_apply_situ_epilogue` in `blackwell/blockscaled_contiguous_gather_grouped_gemm_act_fusion.py:2836-2924`. Copying is defensible if the Rubin kernel must stay self-contained, but a numerics fix in one will not reach the other. Consider lifting the packed-f32x2 sigmoid/tanh helpers into `cute_dsl_kernels/utils` (next to `silu_f32`/`sigmoid_f32`) and calling them from both kernels, or at minimum add a cross-reference comment in both places so the drift risk is visible.

File: tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py [hunk 3]
```diff
@@ -2045,10 +2065,33 @@ def test_nvfp4_kernel_actually_applies_situ(moe_backend):
...
+        from tensorrt_llm._torch.moe.fused_moe.fused_moe_cute_dsl_fc12 import (
+            TrtllmCutedslFusedFc12Nvfp4Impl,
+        )
+
+        assert type(moe.backend) is TrtllmCutedslFusedFc12Nvfp4Impl
     _load_nvfp4_bank_for(moe, bank, moe_backend)
 
     router_logits = gate.compute_logits(x)
     actual = moe.forward(x, router_logits, all_rank_num_tokens=None).float()
+    assert torch.isfinite(actual).all()
```
Comment (brnguyen2) [inline tests/unittest/_torch/moe/test_kimi_k3_situ_moe.py]: This test asserts `situ_l2 < _SITU_NVFP4_REL_L2_MAX` (0.20, line 1836) further down, but the PR description reports rel_l2 0.232667 for `CUTEDSL_FC12`. That run predates the merge of #19003, which introduced the calibrated bound. Because the FC12 parametrization skips on every SM except 107, no CI lane will exercise this; please rerun at head on Rubin and report cosine, rel_l2 and `|k|/|ref|`. If it fails, follow the comment above the bound: 0.233 ≈ sqrt(6)*0.095 vs the other backends' sqrt(3)*0.095, which points at extra quantization stages in the FC12 path rather than at the bound.

File: tests/integration/test_lists/test-db/l0_b200.yml [hunk 4]
```diff
@@ -183,7 +183,7 @@ l0_b200:
   - unittest/_torch/moe/test_moe_backend.py::test_megamoe_streaming_reload_resets_slot_claims
   - unittest/_torch/moe/test_moe_backend.py::test_trtllm_gen_nvfp4_situ_selects_padded_quant_method
   - unittest/_torch/moe/test_moe_backend.py::test_trtllm_gen_nvfp4_situ_fc31_scale_c_drops_dequant_scale
-  - unittest/_torch/moe/test_moe_backend.py::test_megamoe_bakes_situ_softcaps_as_uniform_scalars
+  - unittest/_torch/moe/test_moe_backend.py::test_codegen_baked_situ_softcaps_are_uniform_scalars
```
Comment (brnguyen2) [inline tests/integration/test_lists/test-db/l0_b200.yml]: The rename is complete (only this list had the old name), but the four new GPU-free tests in `test_moe_backend.py` (`test_fc12_rejects_nonuniform_situ_softcaps`, `test_fc12_swiglu_keeps_clamp_without_situ_constants`, `test_fc12_outer_tuning_separates_activation_and_softcaps`, `test_fc12_preparation_primes_fused_tiles_without_two_op_memset_knob`) and `test_moe_impl.py::test_fc12_pinned_activation_support` are not registered here or in any other l0 list, so CI never runs them. This file lists `test_moe_backend.py` per-function, and `test_moe_impl.py` is not in any list at all. Please add the new functions here alongside this entry.

File: tensorrt_llm/_torch/moe/fused_moe/fused_moe_cute_dsl_fc12.py [hunk 5]
```diff
@@ -78,6 +78,21 @@ class CuteDslFc12FusedMoENvfp4Runner(CuteDslFusedMoENvfp4Runner):
     def _tile_sizes():
         return [128, 256]
 
+    def forward(
+        self,
+        inputs: List[torch.Tensor],
+        tactic: Optional[int],
+        do_preparation: bool = False,
+    ) -> torch.Tensor:
+        if do_preparation:
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/moe/fused_moe/fused_moe_cute_dsl_fc12.py]: Worth stating in the docstring that this override exists because setting `workload_identity` (new in this PR) turns on the parent's priming loop, which passes `overlap_moe_output_memset=False`, a kwarg `run_moe_nvfp4_impl` does not accept. Before this PR FC12 preparation was effectively a no-op (identity was `None`), so this is a behavior change for the SwiGLU path too, not just SiTU; the PR description should mention it.

**Review by brnguyen2 — COMMENTED** [review]
The reported Rubin numerics contradict the test at head, and CI cannot catch it. The description says the real `CUTEDSL_FC12` SiTU test passed with relative L2 0.232667 against the SiTU reference, but `_SITU_NVFP4_REL_L2_MAX` in `test_kimi_k3_situ_moe.py` is 0.20 (from #19003, which landed in this branch via the merge from `main`, after the Slurm run on `51ec5e46`). At the current head that number fails the assertion. Since the test skips on every SM except 107, the manual Rubin run is the only merge gate for this path, and it is stale. Please rerun `test_nvfp4_kernel_actually_applies_situ[CUTEDSL_FC12]` at head and post cosine, rel_l2 and `|k|/|ref|`.

Beyond pass/fail, the number itself is suspicious. The bound's comment derives it as sqrt(stages)×0.095; the other three NVFP4 backends measure 0.161–0.166 (sqrt(3)×eps). 0.2327/0.095 ≈ 2.45 ≈ sqrt(6), i.e. roughly three extra quantization-stage-equivalents, and it also exceeds the measured plain-SwiGLU distance (≥0.205). The SiTU epilogue math is copied verbatim from the Blackwell act-fusion kernel that scores 0.16, so the excess is more likely in the FC12-specific path (in-kernel requant/SFC generation, alpha application) than in the activation formula. Comparing FC12 SwiGLU vs two-op CUTEDSL SwiGLU on the same shape would tell whether this is SiTU-specific or an FC12-wide precision characteristic. Please do not loosen the bound without that analysis; the comment above it says why.

Smaller items:
- The description is truncated mid-sentence ("Incorrect SwiGLU ref") — that number is exactly what is needed for the analysis above.
- `[None]` on a feature PR that adds a new kernel epilogue and backend capability: please attach a JIRA.
- The `CuteDslFc12FusedMoENvfp4Runner.forward` / `do_preparation` override is a behavior change (FC12 preparation went from a no-op to priming both tiles) that the description does not mention.
- `CUTEDSL_FC12` is not mentioned anywhere under `docs/source` (pre-existing since #18357). A line in the MoE backend docs listing it and its SwiGLU/SiTU support would help.
